exporter/otelarrowexporter/otelarrow.go

```diff
@@ -7,72 +7,329 @@ import (
+	e.metricExporter = pmetricotlp.NewGRPCClient(e.clientConn)
+	e.logExporter = plogotlp.NewGRPCClient(e.clientConn)
+	headers := map[string]string{}
+	for k, v := range e.config.ClientConfig.Headers {
+		headers[k] = string(v)
+	}
+	e.metadata = metadata.New(headers)
+	e.callOptions = []grpc.CallOption{
+		grpc.WaitForReady(e.config.ClientConfig.WaitForReady),
+	}
+
+	if !e.config.Arrow.Disabled {
+		// Note this sets static outgoing context for all future stream requests.
+		ctx := e.enhanceContext(context.Background())
+
+		var perRPCCreds credentials.PerRPCCredentials
+		if e.config.ClientConfig.Auth != nil {
+			// Get the auth extension, we'll use it to enrich the request context.
+			authClient, err := e.config.ClientConfig.Auth.GetClientAuthenticator(host.GetExtensions())
+			if err != nil {
+				return err
+			}
+
+			perRPCCreds, err = authClient.PerRPCCredentials()
+			if err != nil {
+				return err
+			}
+		}
+
+		arrowOpts := e.config.Arrow.toArrowProducerOptions()
+
+		arrowCallOpts := e.callOptions
+
+		if e.config.ClientConfig.Compression == configcompression.TypeZstd {
+			// ignore the error below b/c Validate() was called
+			_ = zstd.SetEncoderConfig(e.config.Arrow.Zstd)
+			// use the configured compressor.
+			arrowCallOpts = append(arrowCallOpts, e.config.Arrow.Zstd.CallOption())
+		}
+
+		e.arrow = arrow.NewExporter(e.config.Arrow.MaxStreamLifetime, e.config.Arrow.NumStreams, e.config.Arrow.Prioritizer, e.config.Arrow.DisableDowngrade, e.settings.TelemetrySettings, arrowCallOpts, func() arrowRecord.ProducerAPI {
+			return arrowRecord.NewProducerWithOptions(arrowOpts...)
+		}, e.streamClientFactory(e.clientConn), perRPCCreds, e.netReporter)
+
+		if err := e.arrow.Start(ctx); err != nil {
+			return err
+		}
+	}
+
 	return nil
 }
 
-func (e *baseExporter) shutdown(_ context.Context) error {
-	// TODO: Implementation.
+func (e *baseExporter) shutdown(ctx context.Context) error {
+	var err error
+	if e.arrow != nil {
+		err = multierr.Append(err, e.arrow.Shutdown(ctx))
+	}
+	if e.clientConn != nil {
+		err = multierr.Append(err, e.clientConn.Close())
+	}
+	return err
+}
+
+// arrowSendAndWait gets an available stream and tries to send using
+// Arrow if it is configured.  A (false, nil) result indicates for the
+// caller to fall back to ordinary OTLP.
+//
+// Note that ctx is has not had enhanceContext() called, meaning it
+// will have outgoing gRPC metadata only when an upstream processor or
+// receiver placed it there.
+func (e *baseExporter) arrowSendAndWait(ctx context.Context, data any) (sent bool, _ error) {
+	if e.arrow == nil {
+		return false, nil
+	}
+	sent, err := e.arrow.SendAndWait(ctx, data)
+	if err != nil {
+		return sent, processError(err)
+	}
+	return sent, nil
+}
+
+func (e *baseExporter) pushTraces(ctx context.Context, td ptrace.Traces) error {
+	if sent, err := e.arrowSendAndWait(ctx, td); err != nil {
+		return err
+	} else if sent {
+		return nil
+	}
+	req := ptraceotlp.NewExportRequestFromTraces(td)
```

Comment:
Please help me understand - shouldn't we use the `DisableDowngrade` configuration option here? This code seems to always downgrade to OTLP if could not send with arrow. Am I missing something?

Reply:
The design of this component calls for it to fall back to standard OTLP when Arrow is unavailable.
This is why we copied from the core OTLP exporter to begin with (and the receiver, likewise) and earlier in the project we decided to sever the history, as it wasn't helping us. 

If the `DisableDowngrade` option is set, the `arrowSendAndWait()` call above will always return `err != nil` or `sent == true` to avoid falling through. When the Arrow code decides to downgrade, it stops trying and quickly returns `sent=false && err = nil`.

Having implemented all of this, in hindsight, I'm not sure I would go this way again. The downgrade motion can be confusing in its own right, and the code is a lot more complex from having the option to downgrade (and there are some corner cases I'm not happy about). Still, it's designed to make the upgrade painless for users, and I can see how it will help them.
```suggestion
	req := ptraceotlp.NewExportRequestFromTraces(td)and
```